docs: add explanations for how to manage service dependencies

IronCore864 opened the pull request:
A how-to guide on managing service dependencies.

Preview: https://canonical-pebble--440.com.readthedocs.build/en/440/how-to/service-dependencies/

Note that this PR include changes from https://github.com/canonical/pebble/pull/438; to only review the changes on the service dependencies how-to guide, see [here](https://github.com/IronCore864/pebble/pull/2).

benhoyt commented:
Note that I've fixed and updated the wording in the README "service dependencies" section -- we may need to update the content in the docs in parallel (till the docs site is published and the old README content is deleted).

benhoyt commented on docs/how-to/service-dependencies.md:
```diff
@@ -1,11 +1,297 @@
+
+Apparently, the above layer configuration works but is not exactly ideal, because:
+
+- In this example, the three components kind of rely on each other, without anyone, our whole web app won't function properly.
```
```suggestion
- In this example, the three components rely on each other. Without any one of them, our whole web app won't function properly.
```

benhoyt commented on docs/how-to/service-dependencies.md:
```diff
@@ -1,11 +1,297 @@
+Apparently, the above layer configuration works but is not exactly ideal, because:
+
+- In this example, the three components kind of rely on each other, without anyone, our whole web app won't function properly.
+- If the database isn't started (or can't be started for some reason), it doesn't make sense to start the backend that depends on it. Even if the backend starts, it fails to communicate with the database, so the whole web app won't function properly.
```
```suggestion
- If the database isn't started (or can't be started for some reason), it doesn't make sense to start the backend that depends on it. Even if the backend started, it would fail to communicate with the database, so the whole web app won't function properly.
```

benhoyt commented on docs/how-to/service-dependencies.md:
```diff
@@ -1,11 +1,297 @@
+
+- In this example, the three components kind of rely on each other, without anyone, our whole web app won't function properly.
+- If the database isn't started (or can't be started for some reason), it doesn't make sense to start the backend that depends on it. Even if the backend starts, it fails to communicate with the database, so the whole web app won't function properly.
+- The same is true for the frontend: if the backend isn't started (fails to start, can't be started, etc.), it makes no sense to start the frontend alone.
```
```suggestion
- The same is true for the frontend: if the backend isn't started (if it fails to start or can't be started), it makes no sense to start the frontend alone.
```

benhoyt commented on docs/how-to/service-dependencies.md:
```diff
@@ -1,11 +1,297 @@
+- If the database isn't started (or can't be started for some reason), it doesn't make sense to start the backend that depends on it. Even if the backend starts, it fails to communicate with the database, so the whole web app won't function properly.
+- The same is true for the frontend: if the backend isn't started (fails to start, can't be started, etc.), it makes no sense to start the frontend alone.
+
+For example, when the DB fails to start, other services are still up and running:
```
```suggestion
The following example shows the database failing to start, but the database and frontend starting successfully:
```

benhoyt commented on docs/how-to/service-dependencies.md:
````diff
@@ -1,11 +1,297 @@
+frontend-server  enabled  active    today at 10:20 CST
+```
+
+As we can see, although `database` is in `inactive` state, the backend/frontend are still both started, which isn't exactly useful since all they are doing is consuming resources without being able to provide any service.
````
```suggestion
As we can see, although the `database` is `inactive`, the backend and frontend are both `active`. This isn't exactly useful, since all they are doing is consuming resources without being able to provide any service.
```

benhoyt commented on docs/how-to/service-dependencies.md:
```diff
@@ -1,11 +1,297 @@
+
+As we can see, although `database` is in `inactive` state, the backend/frontend are still both started, which isn't exactly useful since all they are doing is consuming resources without being able to provide any service.
+
+Wouldn't it be nice to make these services depend on each other, and when a dependency fails to start, put other services that rely on it on hold, instead of starting them anyway?
```
This is quite a long pile-up of phrases. Maybe this?

```suggestion
Wouldn't it be nice to tell Pebble that these services depend on each other? If a dependency failed to start, Pebble could put the services that rely on it "on hold" instead of starting them. This is where service dependencies come in.
```

benhoyt commented on docs/how-to/service-dependencies.md:
```diff
@@ -1,11 +1,297 @@
+
+Wouldn't it be nice to make these services depend on each other, and when a dependency fails to start, put other services that rely on it on hold, instead of starting them anyway?
+
+Worry no more, service dependency comes to the rescue.
```
I think "worry no more" and "to the rescue" is just a bit too casual / unprofessional. I've rolled it into the last sentence above.

benhoyt commented on docs/how-to/service-dependencies.md:
```diff
@@ -1,11 +1,297 @@
+
+## Service dependencies
+
+Pebble can take service dependencies into account when managing services: the magical `required` (see more detail in [Layer specification](../reference/layer-specification.md)) feature.
```
Again, I think "magical" is a bit much here. I get that you're trying to have a fun tone, but I think it strikes the wrong chord for this kind of documentation. It would be fine in a personal blog post.

Also note that it's `requires`, not `required`.

```suggestion
Pebble can take service dependencies into account when managing services: this is done with the `required` list in the [service definition](../reference/layer-specification.md).
```

benhoyt commented on docs/how-to/service-dependencies.md:
```diff
@@ -1,11 +1,297 @@
+
+Pebble can take service dependencies into account when managing services: the magical `required` (see more detail in [Layer specification](../reference/layer-specification.md)) feature.
+
+Simply put, you can configure a list of other services in the `required` section to indicate this service requires those other services to start correctly.
```
```suggestion
Simply put, you can configure a list of other services in the `requires` section to indicate this service requires those other services to start correctly.
```

benhoyt commented on docs/how-to/service-dependencies.md:
```diff
@@ -1,11 +1,297 @@
+
+Simply put, you can configure a list of other services in the `required` section to indicate this service requires those other services to start correctly.
+
+When Pebble starts a service, it also starts the services which that service depends on (configured with `required`). Conversely, when stopping a service, Pebble also stops services which depend on that service.
```
```suggestion
When Pebble starts a service, it also starts the services which that service depends on (configured with `requires`). Conversely, when stopping a service, Pebble also stops services which depend on that service.
```

benhoyt commented on docs/how-to/service-dependencies.md:
```diff
@@ -1,11 +1,297 @@
-When multiple services need to be started together, they're started in order according to the `before` and `after` configuration, waiting 1 second for each to ensure the command doesn't exit too quickly. The `before` option is a list of services that this service must start before (it may or may not `require` them). Or if it's easier to specify this ordering the other way around, `after` is a lis ...
+## Start order
+
+When multiple services need to be started together, they're started in order according to the `before` and `after` configuration (see more detail in [Layer specification](../reference/layer-specification.md)), waiting for 1 second for each to ensure the command doesn't exit too quickly.
```
I wonder if we can again avoid the parenthesis and just link "configuration" directly:

```suggestion
When multiple services need to be started together, they're started in order according to the `before` and `after` in the [layer configuration](../reference/layer-specification.md). Pebble waits 1 second after starting each service to ensure the command doesn't exit too quickly.
```

benhoyt commented on docs/how-to/service-dependencies.md:
```diff
@@ -1,11 +1,297 @@
+
+## Putting it together
+
+With the service dependencies and start order features, let's reconfigure our web app layer to:
```
```suggestion
With the service dependencies and start order features, let's reconfigure our web app layer:
```

benhoyt commented on docs/how-to/service-dependencies.md:
```diff
@@ -1,11 +1,297 @@
+2024-06-28T02:17:23.347Z [pebble] Started daemon.
+[backend-server database frontend-server]
+2024-06-28T02:17:23.353Z [pebble] POST /v1/services 2.613042ms 202
+2024-06-28T02:17:23.356Z [pebble] Service "backend-server" starting: python3 -m http.server 8081
+2024-06-28T02:17:24.363Z [pebble] Service "database" starting: python3 -m http.server 3306
+2024-06-28T02:17:25.370Z [pebble] Service "frontend-server" starting: python3 -m http.server 8080
```
With the configuration above, I would expect to see database, backend-server, then frontend-server -- in that order due to the `after` configuration. That's what I get locally -- was this a copy 'n' paste error?

benhoyt commented on docs/how-to/service-dependencies.md:
````diff
@@ -1,11 +1,297 @@
+2024-06-28T02:17:26.385Z [pebble] Started default services with change 1.
+```
+
+And we can verify the status of the services:
````
```suggestion
You can see how `database` is started first, then `backend-server`, then `frontend-server` -- this is due to the `after` ordering. And we can verify the status of the services:
```

benhoyt commented on docs/how-to/service-dependencies.md:
````diff
@@ -1,11 +1,297 @@
+frontend-server  enabled  active   today at 10:17 CST
+```
+
+OK, there seems no difference when everything is working. But when one service fails to start, we can see the difference - for example, when the DB can't be started (say, the port is already taken):
````
```suggestion
There seems to be no difference when everything is working. But when one service fails to start, we can see the difference. For example, here's what happens if the database can't be started (say, the port is already taken):
```

benhoyt commented on docs/how-to/service-dependencies.md:
````diff
@@ -1,11 +1,297 @@
+frontend-server  enabled  inactive  -
+```
+
+We can see that since the database failed to start, the backend and the frontend are not started. We can further verify this in Pebble [changes and tasks](../reference/changes-and-tasks/changes-and-tasks.md):
````
```suggestion
We can see that since the database failed to start, the backend and the frontend are not started. We can further verify this using [changes and tasks](../reference/changes-and-tasks/changes-and-tasks.md):
```

benhoyt commented on docs/how-to/service-dependencies.md:
```diff
@@ -1,11 +1,297 @@
+
+We can see that since the database failed to start, the backend and the frontend are not started. We can further verify this in Pebble [changes and tasks](../reference/changes-and-tasks/changes-and-tasks.md):
+
+Change:
```
No need for this paragraph, as we already have a `:` after the paragraph above.

```suggestion
```

benhoyt commented on docs/how-to/service-dependencies.md:
````diff
@@ -1,11 +1,297 @@
+1    Error   today at 10:28 CST  today at 10:28 CST  Autostart service "database" and 2 more
+```
+
+Task:
````
"Task:" as a paragraph doesn't stand by itself. Perhaps just merge the two terminal blocks into one.

benhoyt commented on docs/how-to/service-dependencies.md:
````diff
@@ -1,11 +1,297 @@
+2024-06-28T10:28:06+08:00 ERROR cannot start service: exited quickly with code 1
+```
+
+We can see that since the database failed to start, other services that depend on it are put in "Hold" status.
````
"Hold" is actually a changes & task thing, not a service status. Maybe this:

```suggestion
We can see that since the database failed to start, other services that depend on it are not started (their service status remains "inactive").
```

benhoyt commented on docs/how-to/service-dependencies.md:
```diff
@@ -1,11 +1,297 @@
+
+## See more
+
+- [Pebble services command](../reference/pebble-services.md)
+- [Pebble start command](../reference/pebble-start.md)
+- [Pebble stop command](../reference/pebble-stop.md)
+- [Changes and tasks](../reference/changes-and-tasks/changes-and-tasks.md)
+- [Layer specification](../reference/layer-specification.md)
```
```suggestion
- [`pebble services` command](../reference/pebble-services.md)
- [`pebble start` command](../reference/pebble-start.md)
- [`pebble stop` command](../reference/pebble-stop.md)
- [Changes and tasks](../reference/changes-and-tasks/changes-and-tasks.md)
- [Layer specification](../reference/layer-specification.md)
```

benhoyt commented on docs/how-to/service-dependencies.md:
```diff
@@ -1,11 +1,297 @@
+
+Let's have a look at a concrete example of service dependencies.
+
+## My web app
```
I find "my" a bit too casual, perhaps:

```suggestion
## A simple web application
```

benhoyt reviewed (commented):
Structure and approach looks good to me, but I've left a bunch of comments about style and grammar.

AnneCYH commented on docs/how-to/service-dependencies.md:
```diff
@@ -1,11 +1,297 @@
+
+And when orchestrating more services, things can and will get trickier, because more often than not, those services depend on each other to function together, and managing dependencies is no mean feat.
+
+Let's have a look at a concrete example of service dependencies.
```
I suggest removing this and providing more context instead. Similar to the feedback provided about the tutorial.

E.g.

This document shows how to manage a set of services that have dependencies on each other to function properly. A simple web application is used to demonstrate the instructions in this guide.

AnneCYH commented on docs/how-to/service-dependencies.md:
````diff
@@ -1,11 +1,297 @@
+```{note}
+We use Python's http module to mock the servers.
+
+Alternatively, you may choose to define those services in a single layer. For example:
````
I agree that we don't need an alternative here. If there's a reason to present single YAML vs multiple YAML files for layers, it can be a separate guide / tutorial.

I agree that if a single YAML file does the job, then the sample provided in the beginning can be shown in 1 layer instead of multiple layers.

AnneCYH commented on docs/how-to/service-dependencies.md:
```diff
@@ -1,11 +1,297 @@
+
+## My web app
+
+Imagine we are running a web application, which has:
```
For how-tos I think the language should be more matter-of-fact.
https://diataxis.fr/how-to-guides/#the-language-of-how-to-guides

I suggest rewording this section to something like this.

To demonstrate the instructions in this guide, we are using a simple web
application with the following setup:

- A database listening on port 3306
- A backend server listening on port 8081, which talks to the database
- A frontend server listening on port 8080, which talks to the backend

The relationship between the frontend server, backend server and database can be simplified as:

`frontend (8080) -> backend (8081) -> database (3306)`

IronCore864 replied:
I merged this suggestion with Ben's.

IronCore864 commented:
I have resolved all comments, merged suggestions from both Anne and Ben. See the latest preview [here](https://canonical-pebble--440.com.readthedocs.build/en/440/how-to/service-dependencies/#).

IronCore864 commented:
After discussion, we decided to [merge Anne's PR](https://github.com/canonical/pebble/pull/457) first, then add back the removed sections into explanations, which is what this PR now does.

I added the sections into [Service dependencies](https://canonical-pebble--440.com.readthedocs.build/en/440/explanation/service-dependencies/) and [Service start order](https://canonical-pebble--440.com.readthedocs.build/en/440/explanation/service-start-order/), also updated links in the [how-to doc](https://canonical-pebble--440.com.readthedocs.build/en/440/how-to/service-dependencies/).

Please preview the above links and review this PR @benhoyt @AnneCYH.
